ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,478 @@
+        ValidatorProof calldata proof
+    ) external payable {
+        // for pre-submission, we accept commitments from either the current or next validator set
+        ValidatorSet memory vset;
+        if (validatorSetID == currentValidatorSet.id) {
+            vset = currentValidatorSet;
+        } else if (validatorSetID == nextValidatorSet.id) {
+            vset = nextValidatorSet;
+        } else {
+            revert("Unknown validator set");
+        }
+
+        // Check if merkle proof is valid based on the validatorSetRoot
+        require(
+            isValidatorInSet(vset, proof.addr, proof.index, proof.merkleProof),
+            "invalid validator proof"
+        );
+
+        // Check if validatorSignature is correct, ie. check if it matches
+        // the signature of senderPublicKey on the commitmentHash
+        require(ECDSA.recover(commitmentHash, proof.signature) == proof.addr, "Invalid signature");
+
+        // Check that the bitfield actually contains enough claims to be successful, ie, >= 2/3
+        require(
+            bitfield.countSetBits() >= minimumSignatureThreshold(vset),
+            "Not enough claims"
+        );
+
+        // Accept and save the commitment
+        requests[nextRequestID] = Request(
+            msg.sender,
+            validatorSetID,
+            commitmentHash,
+            bitfield,
+            block.number,
+            vset
+        );
+
+        emit NewRequest(nextRequestID, msg.sender);
+
+        nextRequestID = nextRequestID + 1;
+    }
+
+    /**
+     * @dev Submit a commitment for final verification
+     * @param requestID identifier for the request generated by the initial submission
+     * @param commitment contains the full commitment that was used for the commitmentHash
+     * @param proof a struct containing the data needed to verify all validator signatures
+     */
+    function submitFinal(
+        uint256 requestID,
+        Commitment calldata commitment,
+        ValidatorMultiProof calldata proof
+    ) public {
+        Request storage request = requests[requestID];
+
+        require(commitment.validatorSetID == currentValidatorSet.id);
+
+        verifyCommitment(currentValidatorSet, request, commitment, proof);
+
+        latestMMRRoot = commitment.payload.mmrRootHash;
+        latestBeefyBlock = commitment.blockNumber;
+        emit NewMMRRoot(commitment.payload.mmrRootHash, commitment.blockNumber);
+
+        delete requests[requestID];
+    }
+
+    /**
+     * @dev Submit a commitment and leaf for final verification
+     * @param requestID identifier for the request generated by the initial submission
+     * @param commitment contains the full commitment that was used for the commitmentHash
+     * @param proof a struct containing the data needed to verify all validator signatures
+     * @param leaf an MMR leaf provable using the MMR root in the commitment payload
+     * @param leafProof an MMR leaf proof
+     */
+    function submitFinal(
```

Comment:
should we use different names for these 2 functions? like complete and completeAndRollover ?

Reply:
yeh doesn't really matter